### PR TITLE
[QSP-1 Fix] Move balance to proxy with callbacks

### PR DESCRIPTION
**Pull request by junhaoyap**

### CODE REVIEW

File: contracts/proxy.scilla [hunk 1]
```diff
@@ -109,14 +111,33 @@ transition proxyUpdateBlacklister(newBlacklister : ByStr20)
     send msgs
 end
 
-transition proxyMint(to: ByStr20, value : Uint128 )
+transition proxyMint(to: ByStr20, value : Uint128)
```
Comment (AmritKumar) [inline contracts/proxy.scilla:114]: This looks good to me.

File: contracts/proxy.scilla [hunk 2]
```diff
@@ -109,14 +111,33 @@ transition proxyUpdateBlacklister(newBlacklister : ByStr20)
...
+        | None => zero
+        end;
     msg = {_tag : "mint"; _recipient : current_impl; _amount : zero; to : to;
-           value : value; initiator : _sender};
+           value : value; initiator : _sender; to_bal : to_bal; current_supply : current_supply};
     msgs = one_msg msg;
     send msgs
 end
 
+transition mintCallBack(to: ByStr20, new_to_bal: Uint128, new_supply : Uint128)
```
Comment (AmritKumar) [inline contracts/proxy.scilla:129]: This is good too.

File: contracts/xsgd_contract.scilla [hunk 3]
```diff
@@ -278,20 +276,15 @@ transition mint(to: ByStr20, value : Uint128, initiator : ByStr20)
...
-          new_bal = builtin add value current_val;
-          balances[to] := new_bal;
+          new_supply = builtin add value current_supply;
+          new_to_bal = builtin add value to_bal;
           new_allowed_amount = builtin sub a value;
           minterAllowed[initiator] := new_allowed_amount;
           e = {_eventname : "Minted"; minter : initiator; amount : value; recipient : to};
-          event e
+          event e;
+          msg = {_tag : "mintCallBack"; _recipient : _sender; _amount : Uint128 0; to : to; new_to_bal : new_to_bal; new_supply : new_supply};
```
Comment (AmritKumar) [inline contracts/xsgd_contract.scilla:311]: Replace `Uint128 0` with `zero`.

File: contracts/xsgd_contract.scilla [hunk 4]
```diff
@@ -458,7 +424,7 @@ transition transfer(to : ByStr20, value : Uint128, initiator : ByStr20)
       end
    end
 end
-transition transferFrom (from : ByStr20, to : ByStr20, value : Uint128, initiator : ByStr20)
+transition transferFrom (from : ByStr20, to : ByStr20, value : Uint128, initiator : ByStr20, to_bal : Uint128, from_bal : Uint128)
```
Comment (AmritKumar) [inline contracts/xsgd_contract.scilla:457]: Looks good to me.

File: contracts/proxy.scilla [hunk 5]
```diff
@@ -127,34 +148,120 @@ end
 
 transition proxyTransferFrom (from : ByStr20, to : ByStr20, value : Uint128)
```
Comment (AmritKumar) [inline contracts/proxy.scilla:149]: LGTM.

File: contracts/proxy.scilla [hunk 6]
```diff
@@ -127,34 +148,120 @@ end
...
+        | None => zero
+        end;
     msg = {_tag : "transferFrom"; _recipient : current_impl; _amount : zero;
-           from : from; to : to; value : value; initiator : _sender};
+           from : from; to : to; value : value; initiator : _sender; to_bal : to_bal; from_bal : from_bal};
     msgs = one_msg msg;
     send msgs
 end
 
+transition transferFromCallBack(from : ByStr20, to : ByStr20, new_from_bal : Uint128, new_to_bal : Uint128)
```
Comment (AmritKumar) [inline contracts/proxy.scilla:169]: LGTM

File: contracts/xsgd_contract.scilla [hunk 7]
```diff
@@ -47,9 +47,7 @@ field masterMinter : ByStr20 = init_owner
 field paused : Bool = False
 field blacklister : ByStr20 = init_owner
 field blacklisted : Map ByStr20 Uint128 = Emp ByStr20 Uint128
-field balances : Map ByStr20 Uint128 = Emp ByStr20 Uint128
 field allowed : Map ByStr20 (Map ByStr20 Uint128) = Emp ByStr20 (Map ByStr20 Uint128)
-field totalSupply : Uint128  = zero
 field minters : Map ByStr20 Uint128 = Emp ByStr20 Uint128
```
Comment (jjcnn) [inline contracts/xsgd_contract.scilla:79]: My understanding of this change is that `balances` and `total_supply` is moved to the proxy contract to ensure that data migration is not an issue when updating the token contract.

That raises the question as to why it is not necessary to migrate the other parts of the state? Why is it not necessary to migrate `minters`, `blacklisted` and `allowed`, for instance?

The logical end point for that line of reasoning is of course that you end up with just one contract which then cannot be updated. So I guess the question that needs to be answered is this: Which parts of the logic/state of this application do you want to be able to change, and which parts should be set in stone?

As far as I can tell every possible answer to that question comes with its own set of maintenance problems, so it's a question that needs to be carefully considered. However, not considering it does not mean that the issue goes away - it just means you might inadvertently settle on an answer you aren't happy with.

**Review by jjcnn — COMMENTED** [review]
I have left one comment on the change itself, but I also have a comment on the overall design after this change.

It seems to me that the division of responsibilities between the proxy and the token contract is now roughly such that the proxy handles the actual handling of tokens, whereas the token contract performs the permission checks, i.e., that the users involved have the appropriate permissions to perform the handling in question.

Normally this division would be handled the other way around, i.e., the user would interact with the "guard" contract which checks permissions, and if the permission check fails passes the transaction through to the token handling logic itself.

This suggests that what you really need is the following:
- An immutable token contract, only doing the handling of tokens
- A guard contract which checks user permissions, and passes transaction requests through to the token contract.
- A proxy guard which passes requests through to the guard contract.

Each time the guard contract needs to be updated, a new guard contract is deployed, and the proxy is updated to point to the new guard.

File: contracts/proxy.scilla [hunk 8]
```diff
@@ -127,34 +148,120 @@ end
...
+        balances[initiator] := new_burn_balance;
+        totalSupply := new_supply
+    end
+end
+
 transition proxyLawEnforcementWipingBurn(address : ByStr20)
     current_impl <- implementation;
-    msg = {_tag : "lawEnforcementWipingBurn"; _recipient : current_impl; _amount : zero; address : address; initiator : _sender};
+    current_supply <- totalSupply;
+    get_addr_bal <- balances[_sender];
```
Comment (AmritKumar) [inline contracts/proxy.scilla]: @evesnow91 @junhaoyap Shouldn't this be `balances[address]`?

Comment (AmritKumar) [inline contracts/proxy.scilla]:
```suggestion
    get_addr_bal <- balances[address];
```

**Review by AmritKumar — CHANGES_REQUESTED** [review]

File: contracts/proxy.scilla (hunk 8 above)
Comment (AmritKumar) [reply contracts/proxy.scilla]: I have fixed it.

**Pull request merged**